flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,247 @@
+
+  /**
+   * [[PredictOperation]] for vector types. The result type is a [[LabeledVector]].
+   */
+  implicit def predictValues = {
+    new PredictOperation[KMeans, Vector, LabeledVector] {
+      override def predict(
+        instance: KMeans,
+        predictParameters: ParameterMap,
+        input: DataSet[Vector])
+      : DataSet[LabeledVector] = {
+
+        instance.centroids match {
+          case Some(centroids) => {
+            input.map(new SelectNearestCenterMapper).withBroadcastSet(centroids, CENTROIDS)
+          }
+
+          case None => {
+            throw new RuntimeException("The KMeans model has not been trained. Call first fit" +
+              "before calling the predict operation.")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * [[FitOperation]] which iteratively computes centroids that match the given input DataSet by
+   * adjusting the given initial centroids.
+   */
+  implicit def fitKMeans = {
+    new FitOperation[KMeans, Vector] {
+      override def fit(
+        instance: KMeans,
+        fitParameters: ParameterMap,
+        input: DataSet[Vector])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+
+        val centroids: DataSet[LabeledVector] = resultingParameters.get(InitialCentroids).get
+        val numIterations: Int = resultingParameters.get(NumIterations).get
+
+        val finalCentroids = centroids.iterate(numIterations) { currentCentroids =>
+          val newCentroids: DataSet[LabeledVector] = input
+            .map(new SelectNearestCenterMapper).withBroadcastSet(currentCentroids, CENTROIDS)
+            .map(x => (x.label, x.vector, 1.0)).withForwardedFields("label->_1; vector->_2")
+            .groupBy(x => x._1)
+            .reduce((p1, p2) => (p1._1,(p1._2.asBreeze + p2._2.asBreeze).fromBreeze, p1._3 + p2._3))
+            .withForwardedFields("_1")
+            .map(x => LabeledVector(x._1, (x._2.asBreeze :/ x._3).fromBreeze))
+            .withForwardedFields("_1->label")
+
+          newCentroids
+        }
+
+        instance.centroids = Some(finalCentroids)
+      }
+    }
+  }
+
+}
+
+/**
+ * Converts a given vector into a labeled vector where the label denotes the label of the closest
+ * centroid.
+ */
+@ForwardedFields(Array("*->vector"))
+final class SelectNearestCenterMapper extends RichMapFunction[Vector, LabeledVector] {
+
+  import KMeans._
+
+  private var centroids: Traversable[LabeledVector] = null
+
+  /** Reads the centroid values from a broadcast variable into a collection. */
+  override def open(parameters: Configuration) {
+    centroids = getRuntimeContext.getBroadcastVariable[LabeledVector](CENTROIDS).asScala
+  }
+
+  def map(v: Vector): LabeledVector = {
+    var minDistance: Double = Double.MaxValue
+    var closestCentroidLabel: Double = -1
+    centroids.foreach(centroid => {
+      val distance = EuclideanDistanceMetric().distance(v, centroid.vector)
```

Comment:
It would be nice to make the distance metric configurable by the user, so that metrics other than Euclidean distance can be used,